src/gpu/tsvd/tsvd.cu

```diff
@@ -171,14 +170,30 @@ namespace tsvd
 	 * 4.Explained Variance Ratio
 	 */
 	template<typename T, typename S>
-	void get_tsvd_attr(Matrix<T> &X, Matrix<T> &Q, S _Q, Matrix<T> &w, S _w, S _U, S _explained_variance, S _explained_variance_ratio, params _param, DeviceContext &context){
+	void get_tsvd_attr(Matrix<T> &X, Matrix<T> &Q, S _Q, Matrix<T> &w, S _w, S _U, S _explained_variance, S _explained_variance_ratio, params _param, DeviceContext &context, bool _whiten){
 
 		//Obtain Q^T to obtain vector as row major order
 		Matrix<T>Qt(Q.columns(), Q.rows());
 		transpose(Q, Qt, context); //Needed for calculate_u()
 		Matrix<T>QtTrunc(_param.k, Qt.columns());
 		row_reverse_trunc_q(Qt, QtTrunc, context);
-		QtTrunc.copy_to_host(_Q); //Send to host
+
+		if (_whiten) {
+			auto d_q = QtTrunc.data();
+			auto x_sqrt_row = std::sqrt(X.rows());
+			auto d_sigma = w.data();
+			auto column_size = QtTrunc.rows();
+			auto counting = thrust::make_counting_iterator <int>(0);
+			thrust::for_each(counting, counting+QtTrunc.size(), [=]__device__(int idx){
+				int column = idx/column_size;
+				float sigma = d_sigma[column];
+				float q = d_q[idx];
+				d_q[idx] = (q * x_sqrt_row)/std::sqrt(sigma);
+			} );
+			QtTrunc.copy_to_host(_Q); //Send to host
+		} else {
+			QtTrunc.copy_to_host(_Q); //Send to host
```

Comment:
Don't need this `else`, just do `QtTrunc.copy_to_host(_Q);` always (and remove it from the `if(_whiten)`).